### PR TITLE
Add distributed training, and further improve training quality.

### PR DESCRIPTION
**Pull request by tanzhenyu**
This PR provides distributed training to FasterRCNN, and further improve the quality by having ROIGen take dict inputs.

### CODE REVIEW

**Review by ianstenbit — COMMENTED** [review]
I think it'd be worthwhile to move the training script to 
`examples/training/object_detection/voc` (for consistency with our existing training script).

What do you think?

Then it would also be nice to make lots of these tweakable parameters CLI flags (like the ImageNet classification script) so that we can establish a training history for our OD models including RCNN.

Ideally this script would perform OD training for either RCNN or RetinaNet, but probably for now it would be just RCNN?

**Comment by tanzhenyu** [comment]
> I think it'd be worthwhile to move the training script to `examples/training/object_detection/voc` (for consistency with our existing training script).
> 
> What do you think?
> 
> Then it would also be nice to make lots of these tweakable parameters CLI flags (like the ImageNet classification script) so that we can establish a training history for our OD models including RCNN.
> 
> Ideally this script would perform OD training for either RCNN or RetinaNet, but probably for now it would be just RCNN?

IIUC, each model has its own training history, so "how to upload" is a utility function that can be imported in each training script. For now it's not ready to move yet, given we haven't set up metrics computation. I used COCOMAP which gives weird result (0.24 or something), so there's no solid metric to upload and monitor.

**Comment by bhack** [comment]
> IIUC, each model has its own training history, so "how to upload" is a utility function that can be imported in each training script. For now it's not ready to move yet, given we haven't set up metrics computation. I used COCOMAP which gives weird result (0.24 or something), so there's no solid metric to upload and monitor.

Also how we are going to track the hash of the code that produced the uploaded weights

**Comment by ianstenbit** [comment]
> Also how we are going to track the hash of the code that produced the uploaded weights?

We're using a Git commit hash for classification models:
https://github.com/keras-team/keras-cv/blob/master/examples/training/classification/imagenet/training_history.json

**Comment by bhack** [comment]
> We're using a Git commit hash for classification models:
https://github.com/keras-team/keras-cv/blob/master/examples/training/classification/imagenet/training_history.json

Yes sure  but I meant that users will want to have a definition of network (Kera-cv) version at some point. I don't think that we could navigate troughs git tabs markdown tables so long.

File: keras_cv/models/object_detection/faster_rcnn.py [hunk 1]
```diff
@@ -207,7 +202,7 @@ def __init__(
             layer = tf.keras.layers.Dense(units=fc_dim, activation="relu")
             self.fcs.append(layer)
         self.box_pred = tf.keras.layers.Dense(units=4)
-        self.cls_score = tf.keras.layers.Dense(units=classes + 1)
+        self.cls_score = tf.keras.layers.Dense(units=classes + 1, activation="softmax")
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply keras_cv/models/object_detection/faster_rcnn.py:205]: yeah probably no -- if we don't use softmaxed output, the prediction / decoding part would be quite confusing, users wouldn't be able to know what kind of "score" they're dealing with

File: examples/models/object_detection/faster_rcnn/train.py [hunk 2]
```diff
@@ -410,6 +438,6 @@ def eval_step(examples):
     rpn_cls_metric.reset_state()
     rcnn_reg_metric.reset_state()
     rcnn_cls_metric.reset_state()
-    print("{} finished epoch {}".format(datetime.now(), epoch))
+    print("{} finished epoch {}".format(datetime.now(), epoch), flush=True)
```
[the remark this answers is not in the document]
Comment (tanzhenyu) [reply examples/models/object_detection/faster_rcnn/train.py]: Done.

**Comment by tanzhenyu** [comment]
> > We're using a Git commit hash for classification models:
> > https://github.com/keras-team/keras-cv/blob/master/examples/training/classification/imagenet/training_history.json
> 
> Yes sure but I meant that users will want to have a definition of network (Keras-cv) version at some point. I don't think that we could navigate through git tags for markdown tables so long.

That seems to be a wider keras question (or even TF) to me, maybe we should continue this discussion on keras github?
